### DISCUSSION TITLE
Need a CFG file that works for spacy v3  with webtrf   to train a custom ner

### DISCUSSION THREAD

**Discussion by badri-rutgers** [explosion/spaCy#10064] [Help: Model Advice] [answered]
What is the correct way to create a CFG file spacyv3 for webtrf model to train a custom ner?  I am using . CFG as shwn below but getting 0 accuracy after several epochs:

=========================== Initializing pipeline ===========================
[2022-01-14 14:44:58,990] [INFO] Set up nlp object from config
[2022-01-14 14:44:59,003] [INFO] Pipeline: ['transformer', 'ner']
[2022-01-14 14:44:59,003] [INFO] Resuming training for: ['transformer']
[2022-01-14 14:44:59,009] [INFO] Created vocabulary
[2022-01-14 14:44:59,010] [INFO] Finished initializing nlp object
[2022-01-14 14:44:59,976] [INFO] Initialized pipeline components: ['ner']
✔ Initialized pipeline

============================= Training pipeline =============================
ℹ Pipeline: ['transformer', 'ner']
ℹ Initial learn rate: 0.001
E    #       LOSS TRANS...  LOSS NER  ENTS_F  ENTS_P  ENTS_R  SCORE 
---  ------  -------------  --------  ------  ------  ------  ------
/usr/local/lib/python3.7/dist-packages/torch/autocast_mode.py:141: UserWarning: User provided device_type of 'cuda', but CUDA is not available. Disabling
  warnings.warn('User provided device_type of \'cuda\', but CUDA is not available. Disabling')
  0       0          24.41     72.29    0.00    0.00    0.00    0.00
  2     200       11312.20   7433.77    0.00    0.00    0.00    0.00

[nlp]
lang = "en"
pipeline = ["tok2vec","transformer","ner"]
batch_size = 1000
disabled = []
before_creation = null
after_creation = null
after_pipeline_creation = null
tokenizer = {"@tokenizers":"spacy.Tokenizer.v1"}

[components]
[components.transformer]
source = "en_core_web_trf"

[components.ner]
factory = "ner"
incorrect_spans_key = null
moves = null
scorer = {"@scorers":"spacy.ner_scorer.v1"}
update_with_oracle_cut_size = 100

[components.ner.model]
@architectures = "spacy.TransitionBasedParser.v2"
state_type = "ner"
extra_state_tokens = false
hidden_width = 64
maxout_pieces = 2
use_upper = true
nO = null

[components.ner.model.tok2vec]
@architectures = "spacy.Tok2VecListener.v1"
width = ${components.tok2vec.model.encode.width}
upstream = "*"

[components.tok2vec]
factory = "tok2vec"

[components.tok2vec.model]
@architectures = "spacy.Tok2Vec.v2"

[components.tok2vec.model.embed]
@architectures = "spacy.MultiHashEmbed.v2"
width = ${components.tok2vec.model.encode.width}
attrs = ["ORTH","SHAPE"]
rows = [5000,2500]
include_static_vectors = false

[components.tok2vec.model.encode]
@architectures = "spacy.MaxoutWindowEncoder.v2"
width = 96
depth = 4
window_size = 1
maxout_pieces = 3

**Comment by polm** [comment, marked as answer]
You're using a tok2vec and a transformer in the same pipeline, which is not necessary and probably causing weird things to happen. You should only have either a transformer **or** a tok2vec in a pipeline, not both.

You can source the transformer from the pretrained pipeline, [see here](https://spacy.io/usage/processing-pipelines/#:~:text=instead%20of%20providing%20a%20factory) for how to do that, though I would recommend just training from scratch using a GPU config from [the quickstart](https://spacy.io/usage/training#quickstart).

Also, it looks like you either do not have a GPU or do not have it configured correctly. Note that training Transformers on CPU is possible but extremely slow and not recommended.

**Reply by badri-rutgers** [reply]
after using GPU, the training time for 200 sentences is around 45 mts.    The script in 
https://github.com/explosion/projects/blob/v3/pipelines/ner_demo_update/scripts/create_config.py to generate CFG works.
Here the replace_listener is model.tok2vec.  Can this be used  for webtrf model?   how exactly, if needed, do you add TransformerListener. The Spacyv3.3 document says : You can share a single transformer or other tok2vec model between multiple components by adding a Transformer or Tok2Vec . The Transformer component and TransformerListener layer do the same thing for transformer models.  For webtrf, should I not be replacing Transformer listener instead of tok2vec? If so, how to do it?
Thanks..

**Reply by polm** [reply]
Ah, I didn't realize you were trying to retrain the existing NER model - that wasn't clear from your earlier post in the thread.

If you are using that project, to generate a usable config you just need to change `project.yml` to use `en_core_web_trf` instead of the small pipeline. 

The `model.tok2vec` there uses "tok2vec" in a generic sense that includes Transformers, you don't have to change it.

Also note that if you are adding a new entity type, we usually recommend training a new NER component from scratch rather than updating the existing one. If you do update the existing one, if you don't include examples of entities it already has, it is likely to trigger "catastrophic forgetting" and be unable to recognize them. See [the FAQ](https://github.com/explosion/spaCy/discussions/8226) for details on that.

**Reply by badri-rutgers** [reply]
Thank you so much. Yes, I have used an  entity ruler in the frontend to add new entity based on patterns, So, the training corpus has all standard entities plus the new entity.   Still want to understand the diff between tok2vec listener and transformer listener , when to use which and how do you know what listener is being used in the pipeline etc. Hope there is a tutorial somewhere on an example of a pipeline that uses a transformer listener vs a pipeline that uses tok2vec and explains the diff and use cases.  Thanks a million for your earlier replies.

**Reply by badri-rutgers** [reply]
I think time is ripe to write a book on Grokking Spacy !!!!

**Reply by polm** [reply]
You can read more about listeners at their doc pages, like for [Tok2VecListener](https://spacy.io/api/architectures/#Tok2VecListener). You can read about the Listener architecture [here](https://spacy.io/usage/embeddings-transformers#embedding-layers) in the docs. 

The difference between the TransformerListener and Tok2VecListener is actually pretty simple - you just use the TransformerListener with Transformers, and the Tok2VecListener with other tok2vecs. There's not really tradeoffs or anything, the difference is just to support differences in data manipulation.

It's a bit more high-level, but I would encourage you to take a look at the [spaCy course](https://course.spacy.io/en/) if you haven't done so before.

**Reply by badri-rutgers** [reply]
Thanks as always for you time and patience in answering my queries.  I will read up on  the material you have suggested. Some I have  others I will.   All I want to know is for my ner update pipeline (transformer ...., ner)  with webtrf how come replace_listener(model.tok2vec) seems to work and not replace_listener(TransformerListener).

**Reply by polm** [reply]
Where are you calling `replace_listener`? Whatever it is, `replace_listeners` needs to take an instance, not a class, which is why TransformerListener won't work. 

Also, this can be confusing, but to be really clear - we do use the word "tok2vec" in two slightly different ways. 

1. Any component that converts tokens into vectors. This is not only the Tok2Vec architectures but *also* the Transformers. This is why `model.tok2vec` can actually be a Transformer - the model needs something that does this conversion, and in this sense the Transformers and other architectures are both doing the same thing (though in different ways).
2. Non-transformer architectures that fall under definition 1. This is the meaning relevant for the Tok2VecListener, for example.

**Reply by badri-rutgers** [reply]
I am calling here:
[components.ner]
source = "en_core_web_trf"
replace_listeners = ["model.tok2vec"]

**Reply by badri-rutgers** [reply]
also, if I were to call  replace_listeners for Transformers, what is the correct way to  use
replace_listeners['spacy-transformers.TransformerListener.v1'] ?

**Reply by polm** [reply]
Ah, OK. Note that using something in the config is not exactly the same as a function call. 

When you source a component in the config like that, the value of `replace_listeners` is a reference to a *block*, so it's using "tok2vec" in the generic sense (definition 1) above. With `replace_listeners` in a config you almost always want to use "model.tok2vec", and it should be fine here. (It's true we could be more explicit about this in the docs, so we'll work on that - thanks for bringing it to our attention!)

**Reply by ConsciousUniverse** [reply]
> Any component that converts tokens into vectors. This is not only the Tok2Vec architectures but also the Transformers. This is why model.tok2vec can actually be a Transformer - the model needs something that does this conversion, and in this sense the Transformers and other architectures are both doing the same thing (though in different ways).

@polm: I spent hours trying to figure out a working config because of this. The docs are very confusing here.

For instance, this page [https://spacy.io/api/language#replace_listeners](https://spacy.io/api/language#replace_listeners) states:

> tok2vec_name | Name of the token-to-vector component, typically "tok2vec" or "transformer".

A small clarification to the docs would probably be welcomed by those yet to tread this path ... :-)

**Comment by badri-rutgers** [comment]
I trained using the config as in project ner_demo_update. but I am getting zero entities for custom label.  the custom label appears in the
eval list (TITLE AND CEONAME) but I am getting 0 entities
PERSON        100.00    89.44    94.42
TITLE           0.00     0.00     0.00
ORG            98.21    95.51    96.84
GPE           100.00    97.80    98.89
CARDINAL      100.00    71.43    83.33
DATE           98.55    90.67    94.44
TIME          100.00    98.78    99.39
FAC           100.00    28.57    44.44
MONEY          98.63    83.72    90.57
PERCENT       100.00    90.91    95.24
CEONAME         0.00     0.00     0.00
LOC           100.00   100.00   100.00
ORDINAL       100.00   100.00   100.00
PRODUCT       100.00   100.00   100.00
WORK_OF_ART   100.00   100.00   100.00 LAW             0.00     0.00     0.00

________________________________

**Reply by polm** [reply]
What's the output of `spacy debug data` for your pipeline?

This could happen if either your train or dev data has no instances of your new entities. How did you create your training data?

If your training data looks OK, you could try training a from-scratch model with just your entities rather than doing an update to check if that works. If it does work, then something is wrong with how the update is happening, if it doesn't work, then something is wrong with your data.

**Comment by badri-rutgers** [comment]
when i use a factory ner  in config in the transformer pipeline, apparently (i need to understand more) I have to set annotating_components=['transformer'].
with this I am getting cust NERS recognized.
it has been a great learning experience. your  prompt replies are highly appreciated.

badri



________________________________

**Reply by polm** [reply]
Just a note, but you shouldn't need to put the Transformer in `annotating_components` unless it's frozen.

**Comment by badri-rutgers** [comment]
train from scratch model works but  update does not work. Here is the output of debug data for my pipeline

ValueError: [E955] Can't find table(s) lemma_rules for language 'en' in spacy-lookups-data. Make sure you have the package installed or provide your own lookup tables if no default lookups are available for your language.

=============================== NER (per type) ===============================

                   P        R        F
PERSON         99.80    90.55    94.95
DATE           99.00    93.42    96.13
TIME          100.00    97.39    98.68
ORG            97.57    95.95    96.75
PERCENT       100.00    97.44    98.70
GPE           100.00    97.53    98.75
MONEY          98.06    88.37    92.97
**CEONAME         0.00     0.00     0.00**
**TITLE           0.00     0.00     0.00**
ORDINAL       100.00    90.00    94.74
CARDINAL       98.59    80.00    88.33
FAC           100.00    25.00    40.00
LOC           100.00    66.67    80.00
LAW           100.00   100.00   100.00
QUANTITY      100.00    83.33    90.91
WORK_OF_ART   100.00   100.00   100.00
PRODUCT         0.00     0.00     0.00


________________________________

**Reply by polm** [reply]
That's not what the output of `debug data` looks like? You should run a command like `spacy debug data my-config.cfg` and the output should look something like this:

```
============================ Data file validation ============================
✔ Pipeline can be initialized with data
✔ Corpus is loadable

=============================== Training stats ===============================
Language: en
Training pipeline: tok2vec, ner
1977 training docs
1977 evaluation docs
⚠ 1977 training examples also in evaluation data
⚠ Low number of examples to train a new pipeline (1977)

============================== Vocab & Vectors ==============================
ℹ 88111 total word(s) in the data (8626 unique)
⚠ 12 misaligned tokens in the training data
⚠ 12 misaligned tokens in the dev data
ℹ No word vectors present in the package

========================== Named Entity Recognition ==========================
ℹ 1 label(s)
0 missing value(s) (tokens with '-' label)
✔ Good amount of examples for all labels
✔ Examples without occurrences available for all labels
✔ No entities consisting of or starting/ending with whitespace

================================== Summary ==================================
✔ 5 checks passed
⚠ 4 warnings
```

**Reply by badri-rutgers** [reply]
============================ Data file validation ============================
Traceback (most recent call last):
  File "/usr/lib/python3.7/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/lib/python3.7/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/usr/local/lib/python3.7/dist-packages/spacy/__main__.py", line 4, in <module>
    setup_cli()
  File "/usr/local/lib/python3.7/dist-packages/spacy/cli/_util.py", line 71, in setup_cli
    command(prog_name=COMMAND)
  File "/usr/local/lib/python3.7/dist-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/usr/local/lib/python3.7/dist-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/usr/local/lib/python3.7/dist-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/local/lib/python3.7/dist-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/local/lib/python3.7/dist-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/local/lib/python3.7/dist-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/usr/local/lib/python3.7/dist-packages/typer/main.py", line 500, in wrapper
    return callback(**use_params)  # type: ignore
  File "/usr/local/lib/python3.7/dist-packages/spacy/cli/debug_data.py", line 73, in debug_data_cli
    silent=False,
  File "/usr/local/lib/python3.7/dist-packages/spacy/cli/debug_data.py", line 107, in debug_data
    nlp.initialize(lambda: train_corpus(nlp))
  File "/usr/local/lib/python3.7/dist-packages/spacy/language.py", line 1305, in initialize
    proc.initialize(get_examples, nlp=self, **p_settings)
  File "spacy/pipeline/transition_parser.pyx", line 568, in spacy.pipeline.transition_parser.Parser.initialize
  File "/usr/local/lib/python3.7/dist-packages/thinc/model.py", line 299, in initialize
    self.init(self, X=X, Y=Y)
  File "/usr/local/lib/python3.7/dist-packages/spacy/ml/tb_framework.py", line 45, in init
    model.get_ref("tok2vec").initialize(X=X)
  File "/usr/local/lib/python3.7/dist-packages/thinc/model.py", line 299, in initialize
    self.init(self, X=X, Y=Y)
  File "/usr/local/lib/python3.7/dist-packages/thinc/layers/chain.py", line 88, in init
    layer.initialize(X=curr_input)
  File "/usr/local/lib/python3.7/dist-packages/thinc/model.py", line 299, in initialize
    self.init(self, X=X, Y=Y)
  File "/usr/local/lib/python3.7/dist-packages/thinc/layers/chain.py", line 86, in init
    layer.initialize(X=curr_input, Y=Y)
  File "/usr/local/lib/python3.7/dist-packages/thinc/model.py", line 299, in initialize
    self.init(self, X=X, Y=Y)
  File "/usr/local/lib/python3.7/dist-packages/thinc/layers/chain.py", line 86, in init
    layer.initialize(X=curr_input, Y=Y)
  File "/usr/local/lib/python3.7/dist-packages/thinc/model.py", line 299, in initialize
    self.init(self, X=X, Y=Y)
  File "/usr/local/lib/python3.7/dist-packages/spacy_transformers/layers/transformer_model.py", line 131, in init
    hf_model = huggingface_from_pretrained(name, tok_cfg, trf_cfg)
  File "/usr/local/lib/python3.7/dist-packages/spacy_transformers/layers/transformer_model.py", line 251, in huggingface_from_pretrained
    tokenizer = AutoTokenizer.from_pretrained(str_path, **tok_config)
  File "/usr/local/lib/python3.7/dist-packages/transformers/models/auto/tokenization_auto.py", line 489, in from_pretrained
    pretrained_model_name_or_path, trust_remote_code=trust_remote_code, **kwargs
  File "/usr/local/lib/python3.7/dist-packages/transformers/models/auto/configuration_auto.py", line 582, in from_pretrained
    config_dict, _ = PretrainedConfig.get_config_dict(pretrained_model_name_or_path, **kwargs)
  File "/usr/local/lib/python3.7/dist-packages/transformers/configuration_utils.py", line 556, in get_config_dict
    local_files_only=local_files_only,
  File "/usr/local/lib/python3.7/dist-packages/transformers/configuration_utils.py", line 844, in get_configuration_file
    path_or_repo, revision=revision, use_auth_token=use_auth_token, local_files_only=local_files_only
  File "/usr/local/lib/python3.7/dist-packages/transformers/file_utils.py", line 2103, in get_list_of_files
    return list_repo_files(path_or_repo, revision=revision, token=token)
  File "/usr/local/lib/python3.7/dist-packages/huggingface_hub/hf_api.py", line 885, in list_repo_files
    repo_id, revision=revision, token=token, timeout=timeout
  File "/usr/local/lib/python3.7/dist-packages/huggingface_hub/hf_api.py", line 868, in model_info
    r.raise_for_status()
  File "/usr/local/lib/python3.7/dist-packages/requests/models.py", line 941, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 404 Client Error: Not Found for url: https://huggingface.co/api/models/transformer

**Reply by badri-rutgers** [reply]
I get this when i run python !python -m spacy debug data /configs/config.cfg

**Reply by polm** [reply]
Well that's weird. I'm not really sure what's going on here, but to rule out some things, could you use the [quickstart](https://spacy.io/usage/training#quickstart) to make a CPU NER pipeline and try running `debug data` with that? 

It looks like your 404 error is due to either a configuration problem or maybe a bug related to transformers. Using a non-transformers config will let us check that your data is OK and we can come back to the transformers issue after that.

**Reply by badri-rutgers** [reply]
will I be able to use webtrf model with cpu NER pipeline?

**Reply by polm** [reply]
CPU means not transformers, so no, it'll be different. But I'm suggesting this to clarify data issues, not as a permanent solution.

You could also use a GPU config from the quickstart to use a transformer without training from our pretrained model. This is our recommended approach, especially when adding brand new entities.

**Reply by badri-rutgers** [reply]
yes will do that.  BTW, cfg with NER factory  for  custom-entities works fine. ['transformer', 'tagger', 'parser', 'attribute_ruler', 'lemmatizer', 'cust_ner']    works fine.  NER update is not working.  Also, what is the fundamental difference in NER factory (start from scratch) and NER update? what are the pros and cons?

**Reply by badri-rutgers** [reply]
>>You could also use a GPU config from the quickstart to use a transformer without training from our pretrained model
u mean freeze the transformer?  if so, that's what I am doing

**Reply by badri-rutgers** [reply]
When I run CPU  based web_sm model and execute debug, I get this;
============================ Data file validation ============================
✔ Pipeline can be initialized with data
✔ Corpus is loadable

=============================== Training stats ===============================
Language: en
Training pipeline: tok2vec, tagger, parser, attribute_ruler, lemmatizer, ner
Components from other pipelines: ner
Frozen components: tok2vec, tagger, parser, senter, attribute_ruler, lemmatizer
30872 training docs
30872 evaluation docs
⚠ 7338 training examples also in evaluation data

============================== Vocab & Vectors ==============================
ℹ 723988 total word(s) in the data (14978 unique)
ℹ No word vectors present in the package

========================== Named Entity Recognition ==========================
ℹ 20 label(s)
0 missing value(s) (tokens with '-' label)
⚠ Some model labels are not present in the train data. The model
performance may be degraded for these labels after training: 'LANGUAGE'.
✘ 461 invalid whitespace entity spans
⚠ Low number of examples for label 'WORK_OF_ART' (33)
⚠ Low number of examples for label 'NORP' (5)
⚠ Low number of examples for label 'LAW' (46)
⚠ Low number of examples for label 'PRODUCT' (33)
⚠ Low number of examples for label 'QUANTITY' (18)
⚠ Low number of examples for label 'EVENT' (4)
⚠ Low number of examples for label 'LOC' (15)
✔ Examples without occurrences available for all labels
Entity spans consisting of or starting/ending with whitespace characters are
considered invalid.

=========================== Part-of-speech Tagging ===========================
ℹ 1 label(s) in train data
⚠ Some model labels are not present in the train data. The model
performance may be degraded for these labels after training: 'JJ', 'JJR', 'JJS',
'PDT', 'HYPH', 'MD', 'WDT', 'ADD', 'RBR', 'WP', 'VBD', '''', 'SYM', 'RB', 'VB',
'-LRB-', '$', 'WP$', 'XX', 'CC', 'POS', 'UH', 'FW', 'VBG', 'LS', 'RBS', 'WRB',
'AFX', 'PRP$', 'VBN', ',', 'RP', 'DT', 'NNP', 'VBP', '``', 'PRP', 'NFP', '.',
'NNPS', ':', 'VBZ', 'EX', 'NNS', 'NN', '-RRB-', 'TO', 'IN', 'CD'.

============================= Dependency Parsing =============================
ℹ Found 723988 sentence(s) with an average length of 1.0 words.
ℹ 1 label(s) in train data
ℹ 1 label(s) in projectivized train data

================================== Summary ==================================
✔ 3 checks passed
⚠ 10 warnings
✘ 1 error

**Reply by polm** [reply]
Thanks, that's very helpful. 

Couple of things that jump out at me from that output:

It looks like your training and dev data have a lot of overlap. That's fine for getting started but given your data volume they should be completely separate to test your model properly.

> ✘ 461 invalid whitespace entity spans

This is not a huge number given the size of your training data, but it is unusually large, and these kind of errors are not normal. You are included spaces before or after entities in the entity annotations. Those annotations will be basically unusable. You should look at why that is happening.

For the "Low number of examples" warnings, the model is probably going to forget all those labels because you have basically no data for them.

Your part of speech and dependency data look like they're just invalid and you won't be able to train a useful model. I see that in your real config you are trying to train these components - do you actually want to train these, or do you just want to use the pretrained models? 

To clear things up a little I have two questions:

1. Can you give an example of a sentence with your entities in it?
2. What is the ultimate goal here? You want to have `en_core_web_trf` + NER for your custom entities, right? Are you actually going to use the tagger, parser, lemmatizer etc.?

**Comment by badri-rutgers** [comment]
Update: even after installing spacy_lookups, tokens recognized as new labels is still zero. interesting is that new labels appear in eval output but no tokens  recognized as such. Input samples do have sufficient custom entities

**Comment by badri-rutgers** [comment]
```
#config file used
[paths]
train = null
dev = null
vectors = null
init_tok2vec = null

[system]
gpu_allocator = "pytorch"
seed = 1

[nlp]
lang = "en"
pipeline = ["transformer","tagger","parser","attribute_ruler","lemmatizer","ner"]
disabled = []
before_creation = null
after_creation = null
after_pipeline_creation = null
batch_size = 64
tokenizer = {"@tokenizers":"spacy.Tokenizer.v1"}

[components]

[components.attribute_ruler]
source = "en_core_web_trf"

[components.lemmatizer]
source = "en_core_web_trf"

[components.ner]
source = "en_core_web_trf"
replace_listeners = ["model.tok2vec"]

[components.parser]
source = "en_core_web_trf"

[components.tagger]
source = "en_core_web_trf"

[components.transformer]
source = "en_core_web_trf"

[corpora]

[corpora.dev]
@readers = "spacy.Corpus.v1"
path = ${paths.dev}
gold_preproc = false
max_length = 0
limit = 0
augmenter = null

[corpora.train]
@readers = "spacy.Corpus.v1"
path = ${paths.train}
gold_preproc = false
max_length = 0
limit = 0
augmenter = null

[training]
train_corpus = "corpora.train"
dev_corpus = "corpora.dev"
seed = ${system:seed}
gpu_allocator = ${system:gpu_allocator}
dropout = 0.1
accumulate_gradient = 3
patience = 1000
max_epochs = 0
max_steps = 20000
eval_frequency = 200
frozen_components = ["transformer","tagger","parser","attribute_ruler","lemmatizer"]
before_to_disk = null
annotating_components = []

#[training.batcher]
#@batchers = "spacy.batch_by_padded.v1"
#discard_oversize = true
#get_length = null
#size = 2000
#buffer = 256

[training.batcher]
@batchers = "spacy.batch_by_words.v1"
discard_oversize = false
tolerance = 0.2
get_length = null

[training.batcher.size]
@schedules = "compounding.v1"
start = 100
stop = 1000
compound = 1.001
t = 0.0

[training.logger]
@loggers = "spacy.ConsoleLogger.v1"
progress_bar = false

[training.optimizer]
@optimizers = "Adam.v1"
beta1 = 0.9
beta2 = 0.999
L2_is_weight_decay = true
L2 = 0.01
grad_clip = 1.0
use_averages = true
eps = 0.00000001

[training.optimizer.learn_rate]
@schedules = "warmup_linear.v1"
warmup_steps = 250
total_steps = 20000
initial_rate = 0.00005

[training.score_weights]
tag_acc = 0.16
dep_uas = 0.0
dep_las = 0.16
dep_las_per_type = null
sents_p = null
sents_r = null
sents_f = 0.02
lemma_acc = 0.5
ents_f = 0.16
ents_p = 0.0
ents_r = 0.0
ents_per_type = null
speed = 0.0

[pretraining]

[initialize]
vocab_data = null
vectors = null
init_tok2vec = ${paths.init_tok2vec}
after_init = null
#lookups = null

[initialize.lookups]
@misc = "spacy.LookupsDataLoader.v1"
lang = ${nlp.lang}
tables = ["lexeme_norm"]

[initialize.before_init]
@callbacks = "spacy.copy_from_base_model.v1"
tokenizer = "en_core_web_trf"
vocab = "en_core_web_trf"

[initialize.components]

[initialize.components.ner]

[initialize.components.ner.labels]
@readers = "spacy.read_labels.v1"
path = "corpus/labels/ner.json"
require = false

[initialize.components.parser]

[initialize.components.parser.labels]
@readers = "spacy.read_labels.v1"
path = "corpus/labels/parser.json"
require = false

[initialize.components.tagger]

[initialize.components.tagger.labels]
@readers = "spacy.read_labels.v1"
path = "corpus/labels/tagger.json"
require = false

[initialize.tokenizer]
```

**Comment by badri-rutgers** [comment]
Your part of speech and dependency data look like they're just invalid and you won't be able to train a useful model. I see that in your real config you are trying to train these components - do you actually want to train these, or do you just want to use the pretrained models?

>>>>  NO, I do not want to train POS /DEP.  Where in the CFG does it indicate that I am doing this?  How do I ensure that I use pretrained models?

To clear things up a little I have two questions:

  1.  Can you give an example of a sentence with your entities in it?
  >>>> will do. I have to anon it.
  2.  What is the ultimate goal here? You want to have en_core_web_trf + NER for your custom entities, right?
  3.  >>>>>>>>>>>>>.YES:
  4.   Are you actually going to use the tagger, parser, lemmatizer etc.?
  5.  >>>>>>>>>>NO.   (Components are frozen on the CFG)  I only want to use trf+ NER   for two new entities:


________________________________

**Reply by polm** [reply]
Hey, as a general point, it's kind of hard to follow your replies. Github Discussions have an email integration but it doesn't work that great, so it might be better if you used the web UI, and used the Preview feature to check how your post will look. You could also have a look at the [Markdown guide](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax). 

> NO, I do not want to train POS /DEP.  Where in the CFG does it indicate that I am doing this?  How do I ensure that I use pretrained models?

Well, you have all those components in your training pipeline. They are frozen, so they won't actually be trained currently, they'll just run during evaluation. But since they aren't required for NER anyway you might as well just not have them in your `pipeline` and source them later if you're actually going to use them. As it is, they just make evaluation slower. (The demo script automatically puts them in to show how to do it, but if you're not using them it'd be better to just remove them.)

For training new entities, the way we recommend is **not** to re-use the `en_core_web_trf` model, but instead to just use a quickstart config and train from that. That doesn't mean training from scratch - for smaller models you usually use word vectors, and for Transformers you'll start from roberta-base (or another model if you want). If you also want the pre-trained model's entities, you can combine the models, like in the [double NER demo](https://github.com/explosion/projects/tree/v3/tutorials/ner_double).

While the NER update demo you're using allows updating the pretrained components, in order to keep the other components working it also freezes the Transformer component. The NER component by itself can still be updated, but it will learn much more slowly than if the Transformer wasn't frozen, particularly for new entities. This might explain why you get bad performance, though flat 0 still seems strange. (The one case where updating the existing model is something we'd recommend is if you want one of the existing entities but want to improve coverage for them, like if you found the existing PERSON was bad at finding names from a certain country or something.)

I would recommend you:

1. Figure out the whitespace issues with your data, since that might be serious.
2. Train a standalone NER model using a quickstart config. I would use CPU first just to make sure you get non-zero scores; it should be very quick to test.

**Reply by badri-rutgers** [reply]
thank you so much for your time. Will spend time understanding your recommendations.

**Reply by badri-rutgers** [reply]
as to the whitspace issue, I did check and there are no spaces before quotes but almost all my entities are multi word tokens. For example, {"label": "CEONAME", "pattern": "Donald Trump Jr."},  {"label": "ORGNAME", "pattern": "Windmill Associates, INC"},   would this cause whitespace count. I am using these samples to determine spans for each label via entity ruler

**Reply by polm** [reply]
Normal whitespace in between words is not a problem. The warning from `debug data` only applies to incorrect whitespace, like if the entity is between `[]` brackets in `His name is [ Barack Obama]` (starting space). ~400 instances with your data volume is not a lot but it is still worth looking into.

It sounds like you are using an EntityRuler to create your training data? 

One thing to note about your CEONAME pattern is it looks like it's a subset of PERSON labels. That's going to be hard to learn, especially if you try to learn CEONAME and PERSON at the same time, which may just not work. If you want to get CEO names from documents that sounds like it would be better structured as a relation extraction problem. (We've suggested this pattern before in the past, see [this talk](https://github.com/explosion/talks/blob/master/2018-07-07_Building-New-NLP-Solutions-with-spaCy-and-Prodigy.pdf), though I'm not sure if we have a detailed writeup anywhere.).

One other thing - I see you are using the label `ORGNAME`, but the label in the pretrained model is `ORG`.

**Reply by polm** [reply]
Found a detailed writeup of our suggested process for things that are too fine-grained for simple NER:

https://spacy.io/usage/rule-based-matching#models-rules-pos-dep

**Reply by badri-rutgers** [reply]
Starting with a  blank nlp and ner demo replace  works.  Previously , I was using NER DEMO update. 
Here is the  basic flow:

```
nlp=spacy.blank("en")
corpus=[sample_sentences]
ruler=nlp.add_pipe("entity_ruler",first=True)
patterns=[
{"label": "CUSTNER1", "pattern": "Blah Blah NC, LLC"},
     {"label": "CUSTNER2", "pattern": "Vice President"},
......  #additional patterns
]
ruler.add_patterns(patterns)  
#recognize only new custom entities
for sentence in corpus:
    doc=nlp(sentence)
    entities = []
    for ent in doc.ents:
        entities.append([ent.start_char, ent.end_char, ent.label_])
    TRAIN_DATA.append([sentence, {"entities": entities}])

   convert TRAIN_DATA  to  train.spacy and valid.spacy
used NER_DEMO_REPLACE  project/CFG
   after training, load cust ner model to nlp pipeline
custom_ner=spacy.load(/model-best)
nlp.add_pipe(source=custom_ner)
['transformer', 'tagger', 'parser', 'attribute_ruler', 'lemmatizer', 'ner', 'ner_custname']

sample='text'
doc=nlp(sample)
for  ent in doc.ents:
   print(ent.text,ent.label_)
```

The pipeline recognizes both new custom entities as well as the standard ner entities
